src/ruby/ext/grpc/rb_xds_server_credentials.c

```diff
@@ -0,0 +1,163 @@
+/*
+ *
+ * Copyright 2021 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#include "rb_xds_server_credentials.h"
+
+#include <grpc/grpc.h>
+#include <grpc/grpc_security.h>
+#include <grpc/support/log.h>
+#include <ruby/ruby.h>
+
+#include "rb_grpc.h"
+#include "rb_grpc_imports.generated.h"
+#include "rb_server_credentials.h"
+
+/* grpc_rb_cXdsServerCredentials is the ruby class that proxies
+   grpc_server_credentials. */
+static VALUE grpc_rb_cXdsServerCredentials = Qnil;
+
+/* grpc_rb_xds_server_credentials wraps a grpc_server_credentials.  It provides
+   a peer ruby object, 'mark' to hold references to objects involved in
+   constructing the server credentials. */
+typedef struct grpc_rb_xds_server_credentials {
+  /* Holder of ruby objects involved in constructing the server credentials */
+  VALUE mark;
+  /* The actual server credentials */
+  grpc_server_credentials* wrapped;
+} grpc_rb_xds_server_credentials;
+
+/* Destroys the server credentials instances. */
+static void grpc_rb_xds_server_credentials_free(void* p) {
+  grpc_rb_xds_server_credentials* wrapper = NULL;
+  if (p == NULL) {
+    return;
+  };
+  wrapper = (grpc_rb_xds_server_credentials*)p;
+
+  /* Delete the wrapped object if the mark object is Qnil, which indicates that
+     no other object is the actual owner. */
+  if (wrapper->wrapped != NULL && wrapper->mark == Qnil) {
+    grpc_server_credentials_release(wrapper->wrapped);
+    wrapper->wrapped = NULL;
+  }
+
+  xfree(p);
+}
+
+/* Protects the mark object from GC */
+static void grpc_rb_xds_server_credentials_mark(void* p) {
+  grpc_rb_xds_server_credentials* wrapper = NULL;
+  if (p == NULL) {
+    return;
+  }
+  wrapper = (grpc_rb_xds_server_credentials*)p;
```

Comment:
nit: can we declare `wrapper` on the same line that it's initialized?

Reply:
done